File: iOS/MyAIAgent/MyAIAgent/Models/CRMStore.swift
//
//  CRMStore.swift
//  MyAIAgent
//

import Foundation
import Observation

/// Mock CRM the agent's tools act on — customers, appointments, and the lead
/// queue. In-memory by design: the demo is about the agent orchestrating
/// tools, not about persistence. Lives on the main actor because tool results
/// mutate UI-visible state.
@Observable
final class CRMStore {
    nonisolated struct Customer: Identifiable, Equatable {
        let id: UUID
        var name: String
        var phone: String
        var bike: String
        var lastVisit: String
    }

    nonisolated struct Appointment: Identifiable, Equatable {
        let id: String
        var customerName: String
        var service: String
        var day: String
    }

    private(set) var customers: [Customer]
    private(set) var appointments: [Appointment] = []
    var leads: [Lead]

    private var nextAppointmentNumber = 1042

    init(leads: [Lead] = Lead.samples) {
        self.leads = leads
        self.customers = [
            Customer(id: UUID(), name: "Dana Reyes", phone: "555-0117", bike: "Trek FX 3", lastVisit: "March 2026"),
            Customer(id: UUID(), name: "Priya Natarajan", phone: "555-0198", bike: "Specialized Sirrus", lastVisit: "June 2026"),
        ]
    }

    // MARK: Tool operations

    /// Case-insensitive lookup; returns nil when the customer is unknown.
    func lookupCustomer(named name: String) -> Customer? {
        customers.first { $0.name.localizedCaseInsensitiveContains(name) || name.localizedCaseInsensitiveContains($0.name) }
    }

    /// Books an appointment and returns its confirmation id.
    func bookAppointment(customerName: String, service: String, day: String) -> Appointment {
        let appointment = Appointment(
            id: "A-\(nextAppointmentNumber)",
            customerName: customerName,
            service: service,
            day: day
        )
        nextAppointmentNumber += 1
        appointments.append(appointment)
        return appointment
    }

    /// Marks a lead handled so it drops out of the active queue.
    func markLeadHandled(id: UUID) {
        guard let index = leads.firstIndex(where: { $0.id == id }) else { return }
        leads[index].status = .handled
    }

    func markLeadInProgress(id: UUID) {
        guard let index = leads.firstIndex(where: { $0.id == id }) else { return }
        if leads[index].status == .new { leads[index].status = .inProgress }
    }
}

File: android/app/src/main/java/com/curiousbeagle/android/myaiagent/model/CrmStore.kt
package com.curiousbeagle.android.myaiagent.model

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * Mock CRM the agent's tools act on — customers, appointments, and the lead
 * queue. In-memory by design: the demo is about the agent orchestrating
 * tools, not persistence. StateFlow-backed so Compose observes mutations
 * (the `@Observable` analog). Mirrors iOS `CRMStore`.
 */
class CrmStore(initialLeads: List<Lead> = Lead.samples) {

    data class Customer(
        val name: String,
        val phone: String,
        val bike: String,
        val lastVisit: String,
    )

    data class Appointment(
        val id: String,
        val customerName: String,
        val service: String,
        val day: String,
    )

    val customers = listOf(
        Customer("Dana Reyes", "555-0117", "Trek FX 3", "March 2026"),
        Customer("Priya Natarajan", "555-0198", "Specialized Sirrus", "June 2026"),
    )

    private val _leads = MutableStateFlow(initialLeads)
    val leads: StateFlow<List<Lead>> = _leads.asStateFlow()

    private val _appointments = MutableStateFlow<List<Appointment>>(emptyList())
    val appointments: StateFlow<List<Appointment>> = _appointments.asStateFlow()

    private var nextAppointmentNumber = 1042

    fun lead(id: String): Lead? = _leads.value.firstOrNull { it.id == id }

    /** Case-insensitive lookup; null when the customer is unknown. */
    fun lookupCustomer(name: String): Customer? = customers.firstOrNull {
        it.name.contains(name, ignoreCase = true) || name.contains(it.name, ignoreCase = true)
    }

    /** Books an appointment and returns it with its confirmation id. */
    fun bookAppointment(customerName: String, service: String, day: String): Appointment {
        val appointment = Appointment("A-${nextAppointmentNumber++}", customerName, service, day)
        _appointments.update { it + appointment }
        return appointment
    }

    fun markLeadHandled(id: String) = updateLead(id) { it.copy(status = Lead.Status.HANDLED) }

    fun markLeadInProgress(id: String) = updateLead(id) {
        if (it.status == Lead.Status.NEW) it.copy(status = Lead.Status.IN_PROGRESS) else it
    }

    private fun updateLead(id: String, transform: (Lead) -> Lead) {
        _leads.update { leads -> leads.map { if (it.id == id) transform(it) else it } }
    }
}
